# Day 15:Introduction to Large Language Models

**Name:** Areeba Amjad

**Date:** 21 August 2026

**Task:** Introduction to Large Language Models

# 1. What are Large Language Models (LLMs)?

Large Language Models (LLMs) are advanced Artificial Intelligence models trained on very large amounts of text data. They are designed to understand, process, and generate human-like language.

LLMs are commonly based on the **Transformer architecture**, which uses an **attention mechanism** to understand relationships between words and tokens in a sequence.

## Transformer Architecture

The Transformer architecture was introduced in the research paper **"Attention Is All You Need"**. Unlike older sequence models such as RNNs, Transformers can process many tokens in parallel.

The main components include:

- **Input Embeddings:** Convert tokens into numerical vectors.
- **Positional Encoding:** Provides information about the position of tokens.
- **Self-Attention:** Determines which tokens are important in relation to other tokens.
- **Feed-Forward Networks:** Further process the information.
- **Encoder and Decoder:** Used depending on the Transformer architecture and task.

## Attention Mechanism

Attention allows a model to focus on the most relevant parts of the input when processing a particular token.

For example, in the sentence:

> "The cat sat on the mat because it was tired."

The attention mechanism helps the model understand that **"it"** most likely refers to **"the cat"**.

This allows LLMs to understand relationships between words even when they are far apart in a sentence.

## Pre-training vs Fine-tuning

### Pre-training

Pre-training is the initial stage where an LLM learns general language patterns from a very large dataset.

During pre-training, the model learns:

- Grammar
- Vocabulary
- Language patterns
- General knowledge
- Relationships between words and concepts

### Fine-tuning

Fine-tuning is performed after pre-training. The model is trained further on a smaller, task-specific dataset.

For example, an LLM can be fine-tuned for:

- Customer support
- Medical text classification
- Code generation
- Sentiment analysis
- Question answering

### Summary

Pre-training provides **general language capabilities**, while fine-tuning adapts the model for **specific tasks or domains**. 

# 2. Major LLM Providers and Model Comparison

Several major technology companies provide Large Language Models through APIs and applications.

## OpenAI

OpenAI develops the **GPT (Generative Pre-trained Transformer)** family of models.

GPT models are widely used for:

- Text generation
- Question answering
- Summarization
- Coding
- Translation
- Reasoning
- Chatbots

OpenAI models are available through APIs that allow developers to integrate language-model capabilities into applications.

## Anthropic

Anthropic develops the **Claude** family of AI models.

Claude is designed with a strong focus on:

- Helpful responses
- Safety
- Long-context processing
- Document analysis
- Coding
- Reasoning

Claude models are also available through APIs for application development.

## Google

Google develops the **Gemini** family of models.

Gemini models support multiple types of information and can be used for:

- Text generation
- Question answering
- Coding
- Summarization
- Multimodal tasks
- Data analysis

## General Comparison

| Provider | Model Family | Major Strengths | Common Uses |
|---|---|---|---|
| OpenAI | GPT | General-purpose AI, coding, reasoning | Chatbots, coding, content generation |
| Anthropic | Claude | Long-context tasks, document analysis, safety | Research, writing, coding |
| Google | Gemini | Multimodal capabilities, Google ecosystem | Research, analysis, multimodal applications |

The exact capabilities, context limits, pricing, and performance vary by model version. Therefore, model selection should depend on the requirements of the specific application.

# 3. Key Concepts: Tokens, Context Window, and Temperature

Understanding tokens, context windows, and temperature is important when working with LLM APIs.

## Tokens

A token is a unit of text processed by an LLM.

A token can represent:

- A complete word
- Part of a word
- Punctuation
- Numbers
- Special characters

For example, a sentence is converted into tokens before being processed by the model.

LLM APIs generally calculate input and output usage based on tokens.

## Context Window

The context window is the maximum amount of information that a model can process within a single interaction.

It can include:

- Previous conversation messages
- User prompts
- System instructions
- Documents
- Model-generated responses

A larger context window allows the model to work with longer documents and conversations.

However, the context window is not unlimited. If the amount of input becomes larger than the supported context window, some information may need to be removed or summarized.

## Temperature

Temperature controls the randomness of the model's output.

### Low Temperature

A low temperature generally produces:

- More predictable responses
- More consistent wording
- Less variation

It is useful for tasks such as:

- Classification
- Structured responses
- Factual extraction

### High Temperature

A higher temperature generally produces:

- More varied responses
- More creative wording
- Greater randomness

It can be useful for:

- Creative writing
- Brainstorming
- Generating multiple ideas

### Important Note

Temperature does not directly make a model "more intelligent." It mainly changes the randomness of token selection during generation.

# 4. Important LLM API Parameters

LLM APIs provide parameters that control how the model processes prompts and generates responses.

## max_tokens

`max_tokens` controls the maximum amount of output the model can generate.

A larger value allows longer responses, while a smaller value limits the response length.

> Note: The exact parameter name and behavior can vary between providers and model APIs.

## temperature

`temperature` controls the randomness of generated output.

- Lower value → more predictable output
- Higher value → more varied output

## top_p

`top_p` is a sampling parameter that limits token selection to a probability mass.

For example, a smaller `top_p` restricts the model to a smaller set of likely tokens, while a larger value allows more possibilities.

`temperature` and `top_p` both influence randomness, so they are generally tuned carefully rather than being changed aggressively at the same time.

## frequency_penalty

`frequency_penalty` reduces the likelihood of repeatedly using tokens that have already appeared frequently in the generated text.

It can help reduce repetitive responses.

## presence_penalty

`presence_penalty` encourages the model to introduce new topics or tokens rather than repeatedly using the same ones.

It can be useful when generating diverse content.

## Parameter Summary

| Parameter | Purpose |
|---|---|
| `max_tokens` | Controls maximum output length |
| `temperature` | Controls randomness |
| `top_p` | Controls probability-based token sampling |
| `frequency_penalty` | Reduces repeated token usage |
| `presence_penalty` | Encourages new topics/tokens |

Different LLM providers may use different parameter names, supported ranges, or API behavior. Developers should always check the documentation for the specific model being used.

# 5. Common Use Cases of LLMs

LLMs can be used for many Natural Language Processing (NLP) and software development tasks.

## 1. Text Generation

LLMs can generate human-like text based on a prompt.

Examples:

- Blog posts
- Emails
- Reports
- Product descriptions
- Stories

## 2. Summarization

LLMs can convert long documents into shorter summaries while preserving important information.

Examples:

- Research paper summaries
- Meeting notes
- News summaries
- Document summaries

## 3. Classification

LLMs can classify text into predefined categories.

Examples:

- Sentiment analysis
- Spam detection
- Customer complaint classification
- Topic classification

## 4. Question Answering

LLMs can answer questions using information provided in the prompt or retrieved from external sources.

Examples:

- Customer support assistants
- Educational assistants
- FAQ systems

## 5. Code Generation

LLMs can generate, explain, debug, and modify programming code.

Examples:

- Python code
- SQL queries
- API integrations
- Debugging assistance
- Code documentation

## 6. Translation

LLMs can translate text between different languages while considering context and meaning.

## 7. Conversational AI

LLMs can be used to build chatbots and virtual assistants capable of maintaining a natural conversation.

### Overall

LLMs provide a general-purpose language interface that can be integrated into applications across education, business, software development, customer service, research, and many other fields.

# 6. Limitations of Large Language Models

Although LLMs are powerful, they have several important limitations.

## 1. Hallucinations

An LLM may generate information that sounds convincing but is incorrect or unsupported.

This is known as a **hallucination**.

Therefore, important information should be verified using reliable sources.

## 2. Knowledge Cutoff

A model may have limitations regarding information from events or developments that occurred after its training data or knowledge boundary.

For applications requiring current information, external tools such as search, databases, or retrieval systems may be required.

## 3. Context Limits

Every model has a maximum context window.

If a conversation or document exceeds this limit, the complete information may not be available to the model at once.

## 4. Reasoning Errors

LLMs can make mistakes in:

- Mathematical calculations
- Logical reasoning
- Multi-step problems
- Code
- Factual interpretation

Even when the response appears confident, it may still be incorrect.

## 5. Bias

LLMs learn patterns from their training data. If training data contains biases, the model may reproduce or amplify some of those biases.

## 6. Dependency on Prompt Quality

The quality and clarity of the prompt can significantly affect the output.

Poorly written prompts may result in:

- Ambiguous answers
- Missing information
- Incorrect assumptions
- Unwanted output formats

## Conclusion

LLMs are powerful tools but should not always be treated as completely reliable sources of truth. For high-impact applications, their outputs should be validated, monitored, and combined with reliable data sources or human review.

## Task 1 — OpenAI API Setup & Basic Calls

In [1]:
# Install required libraries

!python3 -m pip install openai python-dotenv --break-system-packages


[notice] A new release of pip is available: 26.1.2 -> 26.2.1
[notice] To update, run: pip3 install --upgrade pip


## API Key Setup

The OpenAI API key should never be written directly inside the Python code.

Create a `.env` file in the project directory and add:

```text
OPENAI_API_KEY=your_api_key_here

In [2]:
import os
from dotenv import load_dotenv
from openai import OpenAI

# Load variables from .env
load_dotenv()

api_key = os.getenv("OPENAI_API_KEY")

if not api_key:
    print("API key not found.")
    print("Please create a .env file and add OPENAI_API_KEY.")
else:
    print("API key loaded successfully.")

API key loaded successfully.


In [3]:
from openai import OpenAI

client = OpenAI(api_key=api_key)

print("OpenAI client created successfully.")

OpenAI client created successfully.


In [4]:
# First simple API call

response = client.responses.create(
    model="gpt-5.6-luna",
    input="Explain artificial intelligence in two simple sentences."
)

print("Response:")
print(response.output_text)

RateLimitError: Error code: 429 - {'error': {'message': 'You exceeded your current quota, please check your plan and billing details. For more information on this error, read the docs: https://platform.openai.com/docs/guides/error-codes/api-errors.', 'type': 'insufficient_quota', 'param': None, 'code': 'insufficient_quota'}}

In [5]:
# Task 1: First OpenAI API Call
# Live API execution is unavailable because the account has no API quota.

from openai import RateLimitError

try:
    response = client.responses.create(
        model="gpt-5.6-luna",
        input="Explain artificial intelligence in two simple sentences."
    )

    print("Response:")
    print(response.output_text)

    print("\nToken Usage:")
    print("Input tokens:", response.usage.input_tokens)
    print("Output tokens:", response.usage.output_tokens)
    print("Total tokens:", response.usage.total_tokens)

except RateLimitError as error:
    print("OpenAI API request could not be completed.")
    print("Error type: RateLimitError")
    print("Reason: Insufficient API quota.")
    print("Live API testing is unavailable without available API quota.")

OpenAI API request could not be completed.
Error type: RateLimitError
Reason: Insufficient API quota.
Live API testing is unavailable without available API quota.


## Task 1 — API Execution Limitation

The OpenAI API environment was successfully configured using the `openai` and `python-dotenv` libraries.

The API request was tested using the OpenAI Responses API. However, the request returned a `429 RateLimitError` with the error code `insufficient_quota`.

This indicates that the API account/project currently does not have available API quota for making the request.

Therefore, live API response generation and token usage observation could not be completed without enabling API usage/billing.

The limitation was documented instead of adding or exposing any API key or billing information.

**Security note:** The API key is stored in `.env` and `.env` is included in `.gitignore`. The secret key is not committed to GitHub.

## Task 2 — Parameter Exploration Notebook

In [6]:
# Task 2: Parameter Exploration
# Offline simulation because live OpenAI API quota is unavailable.

prompt = "Explain why Python is useful for Data Science."

temperatures = [0.0, 0.7, 1.5]

simulated_outputs = {
    0.0: (
        "Python is useful for Data Science because it provides "
        "powerful libraries for data analysis, visualization, "
        "machine learning, and automation."
    ),
    0.7: (
        "Python is widely used in Data Science because it is easy "
        "to learn and provides libraries such as Pandas, NumPy, "
        "Matplotlib, and Scikit-learn."
    ),
    1.5: (
        "Python makes Data Science exciting! Its simple syntax, "
        "huge ecosystem, and powerful AI and machine-learning "
        "libraries help data scientists turn raw data into useful insights."
    )
}

for temperature in temperatures:
    print("=" * 60)
    print(f"Temperature: {temperature}")
    print("=" * 60)
    print(simulated_outputs[temperature])

Temperature: 0.0
Python is useful for Data Science because it provides powerful libraries for data analysis, visualization, machine learning, and automation.
Temperature: 0.7
Python is widely used in Data Science because it is easy to learn and provides libraries such as Pandas, NumPy, Matplotlib, and Scikit-learn.
Temperature: 1.5
Python makes Data Science exciting! Its simple syntax, huge ecosystem, and powerful AI and machine-learning libraries help data scientists turn raw data into useful insights.


In [7]:
# Compare temperature effects

temperature_comparison = {
    0.0: "Predictable, consistent, and focused responses.",
    0.7: "Balanced responses with some variation and creativity.",
    1.5: "More varied and creative responses."
}

for temperature, effect in temperature_comparison.items():
    print(f"Temperature {temperature}: {effect}")

Temperature 0.0: Predictable, consistent, and focused responses.
Temperature 0.7: Balanced responses with some variation and creativity.
Temperature 1.5: More varied and creative responses.


In [8]:
# Simulate different output limits

sample_response = (
    "Large Language Models are AI systems trained on large amounts "
    "of text data. They use Transformer architectures and attention "
    "mechanisms to understand relationships between tokens. "
    "LLMs can be used for text generation, summarization, coding, "
    "translation, classification, and question answering."
)

limits = [20, 50, 100]

for limit in limits:
    words = sample_response.split()

    # Simulated truncation
    simulated_output = " ".join(words[:limit])

    print("=" * 60)
    print(f"Maximum output limit: {limit}")
    print("=" * 60)
    print(simulated_output)
    print()

Maximum output limit: 20
Large Language Models are AI systems trained on large amounts of text data. They use Transformer architectures and attention mechanisms

Maximum output limit: 50
Large Language Models are AI systems trained on large amounts of text data. They use Transformer architectures and attention mechanisms to understand relationships between tokens. LLMs can be used for text generation, summarization, coding, translation, classification, and question answering.

Maximum output limit: 100
Large Language Models are AI systems trained on large amounts of text data. They use Transformer architectures and attention mechanisms to understand relationships between tokens. LLMs can be used for text generation, summarization, coding, translation, classification, and question answering.



## max_output_tokens Observation

In an actual LLM API, the maximum output-token parameter controls how much text the model can generate.

A smaller limit can cause the response to stop earlier, while a larger limit allows a longer response.

In this offline experiment, truncation was simulated by limiting the number of words displayed.

| Limit | Expected Effect |
|---|---|
| Small | Short response / possible truncation |
| Medium | Moderate response |
| Large | Longer response |

In [9]:
# Simulate top_p behavior

top_p_results = {
    0.1: [
        "Python is useful for data analysis.",
        "Python has many Data Science libraries.",
        "Python is easy to learn."
    ],
    
    0.9: [
        "Python helps data scientists analyze large datasets.",
        "Python provides an extensive ecosystem for machine learning.",
        "Python can be used to build creative AI and Data Science projects."
    ]
}

for top_p, results in top_p_results.items():
    print("=" * 60)
    print(f"top_p: {top_p}")
    print("=" * 60)

    for result in results:
        print("-", result)

    print()

top_p: 0.1
- Python is useful for data analysis.
- Python has many Data Science libraries.
- Python is easy to learn.

top_p: 0.9
- Python helps data scientists analyze large datasets.
- Python provides an extensive ecosystem for machine learning.
- Python can be used to build creative AI and Data Science projects.



In [10]:
# Simulate top_p behavior

top_p_results = {
    0.1: [
        "Python is useful for data analysis.",
        "Python has many Data Science libraries.",
        "Python is easy to learn."
    ],
    
    0.9: [
        "Python helps data scientists analyze large datasets.",
        "Python provides an extensive ecosystem for machine learning.",
        "Python can be used to build creative AI and Data Science projects."
    ]
}

for top_p, results in top_p_results.items():
    print("=" * 60)
    print(f"top_p: {top_p}")
    print("=" * 60)

    for result in results:
        print("-", result)

    print()

top_p: 0.1
- Python is useful for data analysis.
- Python has many Data Science libraries.
- Python is easy to learn.

top_p: 0.9
- Python helps data scientists analyze large datasets.
- Python provides an extensive ecosystem for machine learning.
- Python can be used to build creative AI and Data Science projects.



# Task 2 Conclusion

The parameter exploration demonstrated how different LLM parameters can influence generated responses.

Key observations:

- Lower temperature generally produces more predictable output.
- Higher temperature can produce more varied output.
- Smaller output limits can cause responses to end earlier.
- Larger output limits allow longer responses.
- Lower `top_p` produces more focused sampling.
- Higher `top_p` allows greater diversity.

Because the OpenAI API account currently has insufficient quota, the experiments were demonstrated using offline simulated outputs rather than live API calls.

## Task 3 — Build Simple Chatbot

In [12]:
# Task 3: Offline Simple Chatbot
# No API or billing required.

print("=" * 50)
print("       OFFLINE AI CHATBOT")
print("=" * 50)
print("Type 'exit' to stop the chatbot.")

while True:
    user_input = input("\nYou: ")

    if user_input.lower() == "exit":
        print("Chatbot: Goodbye!")
        break

    if not user_input.strip():
        print("Chatbot: Please enter a message.")
        continue

    # Simple offline responses
    user_lower = user_input.lower()

    if "python" in user_lower:
        response = (
            "Python is a popular programming language used in "
            "Data Science, AI, web development, and automation."
        )

    elif "llm" in user_lower:
        response = (
            "LLM stands for Large Language Model. "
            "LLMs are AI models trained on large amounts of text."
        )

    elif "hello" in user_lower or "hi" in user_lower:
        response = "Hello! How can I help you today?"

    elif "data science" in user_lower:
        response = (
            "Data Science combines statistics, programming, "
            "machine learning, and data analysis."
        )

    else:
        response = (
            "I am an offline demonstration chatbot. "
            "I can answer basic questions about Python, LLMs, "
            "and Data Science."
        )

    print("Chatbot:", response)

       OFFLINE AI CHATBOT
Type 'exit' to stop the chatbot.



You:  What is python ?


Chatbot: Python is a popular programming language used in Data Science, AI, web development, and automation.



You:  What is LLM?


Chatbot: LLM stands for Large Language Model. LLMs are AI models trained on large amounts of text.



You:  What is data science ?


Chatbot: Data Science combines statistics, programming, machine learning, and data analysis.



You:  exit


Chatbot: Goodbye!


In [13]:
# Define chatbot personality

system_message = """
You are a friendly and beginner-friendly Python and Data Science tutor.
You explain technical concepts using simple language and short examples.
You encourage students while helping them understand programming concepts.
"""

print("Chatbot Role:")
print(system_message)

Chatbot Role:

You are a friendly and beginner-friendly Python and Data Science tutor.
You explain technical concepts using simple language and short examples.
You encourage students while helping them understand programming concepts.



In [14]:
# Offline chatbot with conversation history

conversation_history = []

print("=" * 50)
print("     CONTEXT-AWARE OFFLINE CHATBOT")
print("=" * 50)
print("Type 'exit' to stop.")

while True:
    user_input = input("\nYou: ")

    if user_input.lower() == "exit":
        print("Chatbot: Goodbye!")
        break

    if not user_input.strip():
        print("Chatbot: Please enter a message.")
        continue

    # Store user message
    conversation_history.append({
        "role": "user",
        "content": user_input
    })

    # Generate offline response
    user_lower = user_input.lower()

    if "python" in user_lower:
        assistant_reply = (
            "Python is widely used in Data Science because "
            "of libraries such as NumPy, Pandas and Scikit-learn."
        )

    elif "llm" in user_lower:
        assistant_reply = (
            "LLMs are Large Language Models that process "
            "and generate human-like text."
        )

    elif "what did i ask" in user_lower:
        previous_questions = [
            message["content"]
            for message in conversation_history[:-1]
            if message["role"] == "user"
        ]

        if previous_questions:
            assistant_reply = (
                "Earlier you asked: "
                + " | ".join(previous_questions)
            )
        else:
            assistant_reply = "This is your first question."

    else:
        assistant_reply = (
            "I understand your question. "
            "This is an offline chatbot demonstration."
        )

    # Store assistant response
    conversation_history.append({
        "role": "assistant",
        "content": assistant_reply
    })

    print("Chatbot:", assistant_reply)

     CONTEXT-AWARE OFFLINE CHATBOT
Type 'exit' to stop.



You:  What is dsa?


Chatbot: I understand your question. This is an offline chatbot demonstration.



You:  What is python?


Chatbot: Python is widely used in Data Science because of libraries such as NumPy, Pandas and Scikit-learn.



You:  what is DSA?


Chatbot: I understand your question. This is an offline chatbot demonstration.



You:  what did i ask ?


Chatbot: Earlier you asked: What is dsa? | What is python? | what is DSA?



You:  exit


Chatbot: Goodbye!


In [15]:
# Error handling demonstration

def safe_chatbot_response(user_input):
    try:
        if not isinstance(user_input, str):
            raise TypeError("Input must be text.")

        if not user_input.strip():
            return "Please enter a valid message."

        return "Your message was received successfully."

    except TypeError as error:
        return f"Input Error: {error}"

    except Exception as error:
        return f"Unexpected Error: {error}"


# Test the function
print(safe_chatbot_response("Hello"))
print(safe_chatbot_response(""))

Your message was received successfully.
Please enter a valid message.


# API Error Handling

The original live chatbot was tested with the OpenAI API.

The API returned:

`429 RateLimitError - insufficient_quota`

This means that the API account currently has insufficient quota for live API requests.

The chatbot therefore includes error-handling logic for:

- Rate limits
- API errors
- Connection errors
- Empty user input
- Invalid input
- Unexpected errors

No billing information or API key was added to the project.

In [17]:
# Display conversation history

print("Conversation History")
print("=" * 50)

for message in conversation_history:
    role = message["role"].capitalize()
    content = message["content"]

    print(f"{role}: {content}")

Conversation History
User: What is dsa?
Assistant: I understand your question. This is an offline chatbot demonstration.
User: What is python?
Assistant: Python is widely used in Data Science because of libraries such as NumPy, Pandas and Scikit-learn.
User: what is DSA?
Assistant: I understand your question. This is an offline chatbot demonstration.
User: what did i ask ?
Assistant: Earlier you asked: What is dsa? | What is python? | what is DSA?


# Task 3 Conclusion

A simple interactive chatbot was implemented without requiring paid API access.

The chatbot demonstrates the main concepts required for an LLM chatbot:

- Interactive user input
- Chatbot personality/system instructions
- Conversation history
- Context maintenance
- Error handling
- Input validation
- Response generation

The live OpenAI API version was also tested, but the API returned a `429 insufficient_quota` error. Therefore, an offline implementation was used to demonstrate the chatbot functionality without requiring billing.

# Day 15 — Task Summary

## Task 1: OpenAI API Setup

- Installed `openai` and `python-dotenv`
- Configured `.env`
- Protected the API key using `.gitignore`
- Created an OpenAI client
- Tested a basic API call
- Observed a `429 insufficient_quota` error
- Documented the API limitation

## Task 2: Parameter Exploration

Explored:

- Temperature
- Maximum output tokens
- Top-p
- Frequency penalty
- Presence penalty

Offline simulations were used because live API quota was unavailable.

## Task 3: Simple Chatbot

Implemented:

- Interactive chatbot loop
- Chatbot personality
- Conversation history
- Context maintenance
- Input validation
- Error handling
- Offline response generation

## Billing

Billing and cost calculation were intentionally excluded because the project was completed without paid API usage.

## Overall Learning

This task provided practical understanding of how LLM APIs work, how generation parameters affect model behavior, and how a conversational chatbot can maintain context and handle errors.